src/unique_impl.rs

```diff
@@ -56,7 +56,7 @@ impl<I, V, F> Iterator for UniqueBy<I, V, F>
     type Item = I::Item;
 
     fn next(&mut self) -> Option<Self::Item> {
-        while let Some(v) = self.iter.next() {
+        for v in self.iter.by_ref() {
```

Comment:
Beware that `by_ref` is currently an optimization firewall, so you might not want this change.

See https://github.com/rust-lang/rust/pull/95602 for an example.

Since you want something `&mut self` (like `next` was) here, you could again consider moving the loop into the iterator method.  For example, I think you could phrase this as a `self.iter.find_map(...)` instead for the `for` loop.